src/runtime/hexagon_dma_device_interface.cpp

```diff
@@ -0,0 +1,251 @@
+    halide_hexagon_acquire_dma_context(user_ctxt, &context, num_of_frames);
+    halide_assert(user_context, context != NULL);
+}
+
+HexagonDmaContext::~HexagonDmaContext() {
+    user_ctxt = 0;
+    context = 0;
+}
+}}}}
+
+using namespace Halide::Runtime::Internal::Dma;
+
+extern "C" {
+
+extern WEAK halide_device_interface_t hexagon_dma_device_interface;
+
+WEAK int halide_hexagon_dma_device_release(void *user_context) {
+    return HEX_SUCCESS;
+}
+
+WEAK int halide_hexagon_dma_device_malloc(void *user_context, halide_buffer_t *buf) {
+
+    //////////////////////////////////////////////////////////////////////
+    p_dma_context dma_handle = reinterpret_cast<p_dma_context>(buf->device);
+    uintptr_t  frame = halide_hexagon_dmart_get_frame(user_context, dma_handle);
+    //////////////////////////////////////////////////////////////////////////
+
+    //Check for dma_allocation
+    bool dma_allocate;
+    void* handle = 0;
+    if (halide_hexagon_dmart_allocate_dma(user_context, pdma_context, frame, &dma_allocate) == -1) {
+        error(user_context) << "Undefined Error";
+        return HEX_ERROR;
+    }
+
+    if (dma_allocate) {
+        handle = dma_allocate_dma_engine();
+        if (handle == 0) {
+            error(user_context) << "halide_hexagon_dma_device_malloc:Failed to allocate the read DMA engine.\n";
+            return HEX_ERROR;
+        }
+
+        if (halide_hexagon_dmart_set_dma_handle(user_context, pdma_context, handle, frame)) {
+            error(user_context) << "halide_hexagon_dma_device_malloc:Function failed to set DMA Handle to DMA context \n";
+            return HEX_ERROR;
+        }
+    }
+    return 0;
+}
+
+WEAK int halide_hexagon_dma_device_free(void *user_context, halide_buffer_t *buf) {
+    halide_assert(NULL, buf!=NULL);
+    HexagonDmaContext hexagon_dma_context(user_context);
+    halide_assert(user_context, hexagon_dma_context.get_context() != NULL);
+    p_dma_context dma_ctxt = hexagon_dma_context.get_context();
+
+    halide_hexagon_dmart_delete_context(user_context, dma_ctxt);
+    return HEX_SUCCESS;
+}
+
+WEAK int halide_hexagon_dma_copy_to_device(void *user_context,  halide_buffer_t *buf) {
+    void* handle;
+
+    //////////////////////////////////////////////////////////////////////
+    p_dma_context dma_handle = reinterpret_cast<p_dma_context>(buf->device);
+    uintptr_t  frame = halide_hexagon_dmart_get_frame(user_context, dma_handle);
+    //////////////////////////////////////////////////////////////////////////
+
+    handle = halide_hexagon_dmart_get_dma_handle(user_context, dma_handle, frame);
+    if (handle == 0) {
+        error(user_context) << "Function failed to get DMA Write  Handle  \n";
+        return HEX_ERROR;
+    }
+
+    int ncomponents;
+    ncomponents = halide_hexagon_dmart_get_num_components(user_context, dma_handle, frame);
+
+    t_dma_move_params move_param;
+
+    uintptr_t fold_addr = halide_hexagon_dmart_get_fold_addr(user_context, dma_handle, frame);
```

Comment:
Is this going to be buf->host? It needs to be for this function to work correctly.